File: app/src/lib/game/computer/modules/action-awareness.ts
import type { Candidate, ModuleContext, ScoringModule } from '../types.js';

// Bias toward playing action cards over neutral ones.
// 5 (pick three) is valued most since it hurts the opponent hardest.
// Scores must dominate hand-thinning (max ~12) so action cards always win
// over non-action cards when action-awareness is active.
// Higher modules act as tiebreakers within this priority, not competitors.
const ACTION_SCORES: Partial<Record<number, number>> = {
	5: 30,  // pick three
	2: 25,  // pick two
	14: 20, // general market
	1: 15,  // hold on
	8: 15   // suspension
};

export const actionAwareness: ScoringModule = (candidate: Candidate, _ctx: ModuleContext): number => {
	if (candidate.kind !== 'play-suit') return 0;
	return ACTION_SCORES[candidate.card.value] ?? 0;
};

File: server/src/game/ai/modules/action_awareness.rs
use crate::game::ai::types::{Candidate, ModuleContext};

/// Bias toward action cards. 5 (pick three) valued most; scores dominate hand-thinning.
pub fn action_awareness(candidate: &Candidate, _ctx: &ModuleContext<'_>) -> f64 {
    let Candidate::PlaySuit { value, .. } = candidate else {
        return 0.0;
    };
    match value {
        5 => 30.0,  // pick three
        2 => 25.0,  // pick two
        14 => 20.0, // general market
        1 => 15.0,  // hold on
        8 => 15.0,  // suspension
        _ => 0.0,
    }
}
